Return mapping optimization in tensor_mechanics

WilkAndy opened the pull request:
Rachel and Andy work - long time coming, but it's finally OK.  Thanks @waxmanr ,

a

WilkAndy committed:
Rachel and Andy work on improving tensor_mechanics plasticity performance

TensorMechanicsPlasticModel now has returnMap and consistentTangentOperator methods
that any derived class can override.  An example is TensorMechanicsPlasticJ2

ComputeMultiPlasticityStress now uses these custom returnMap and consistentTangentOperator
methods if possible to dramatically speed up the return-map process.  If the custom
versions fail, or are not coded, or if more than one model is plastically active,
the return-map algorithm falls back to the previous algorithm.

MultiPlasticityRawComponentAssembler has a returnMapAll function that tries all
the custom returnMap functions.

Fixes #5667

WilkAndy committed:
Tests of Rachel and Andy work.

Most of these are just re-goldings due to zero iterations now being required.
The j2_plasticity_vs_LSH stuff now does a CSVDiff instead, and runs for end_time=0.5

Fixes #5667

permcody commented:
@dschwen - You should probably review

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticJ2.h:
```diff
@@ -89,8 +89,95 @@ class TensorMechanicsPlasticJ2 : public TensorMechanicsPlasticModel
+    */
+  bool returnMap(const RankTwoTensor & trial_stress, const Real & intnl_old, const RankFourTensor & E_ijkl, Real ep_plastic_tolerance,
+                        RankTwoTensor & returned_stress, Real & returned_intnl, std::vector<Real> & dpm, RankTwoTensor & delta_dp,
+                        std::vector<Real> & yf, unsigned & trial_stress_inadmissible) const;
```
Any reason `trial_stress_inadmissible` is not a `bool`?

WilkAndy replied:
changed trial_stress_inadmissible to a bool

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticModel.h:
```diff
@@ -138,7 +138,8 @@ class TensorMechanicsPlasticModel : public GeneralUserObject
+  virtual void activeConstraints(const std::vector<Real> & f, const RankTwoTensor & stress, const Real & intnl,
+                                  const RankFourTensor & Eijkl, std::vector<bool> & act, RankTwoTensor & returned_stress) const;
 
   /// Returns the model name (eg "MohrCoulom")
```
typo (MohrCoulomb)

WilkAndy replied:
fixed MohrCoulom typo

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C:
```diff
@@ -125,8 +128,8 @@ ComputeMultiPlasticityStress::computeQpStress()
   if (_fspb_debug == 3)
```
What is `3`? Could an enum be clearer here?

WilkAndy replied:
yea, enum is better, thanks - that's quite old code that i hadn't looked at in months

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C:
```diff
@@ -137,14 +140,20 @@ ComputeMultiPlasticityStress::computeQpStress()
+  cumulative_pm.assign(_num_surfaces, 0);
+
+  // try a "quick" return first - this can be purely elastic, or a customised plastic return defined by a TensorMechanicsPlasticXXXX UserObject
+  bool found_solution = quickStep(_stress_old[_qp], _stress[_qp], _intnl_old[_qp], _intnl[_qp], pm, cumulative_pm, _plastic_strain_old[_qp], _plastic_strain[_qp], _my_elasticity_tensor, _my_strain_increment, _yf[_qp], number_iterations, _Jacobian_mult[_qp], "computeQpStress", true);
```
`found_solution` should be `const` here (the quickStep call might as well be in the if condition)

WilkAndy replied:
i'm not sure what you mean here, @dschwen .  Do you think the code should read

if (!quickStep(blah,blah))
   plasticStep(blah, blah)

?

dschwen replied:
No, the additional bool variable helps the readability, but you should make it `const`. It is just a placeholder that is not intended to be changed.

WilkAndy replied:
aha, i get it.  made found_solution const.  nice one, i hadn't thought of doing that, so you can tell i haven't received any formal training in C.  i wonder how many millions of other places i could do this in my code!

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C:
```diff
@@ -137,14 +140,20 @@ ComputeMultiPlasticityStress::computeQpStress()
-  bool found_solution = elasticStep(_stress_old[_qp], _stress[_qp], _intnl_old[_qp], _intnl[_qp], _plastic_strain_old[_qp], _plastic_strain[_qp], _my_elasticity_tensor, _my_strain_increment, _yf[_qp], number_iterations, _Jacobian_mult[_qp]);
+  // plastic multipliers and cumulative plastic multipliers.
+  // Here they are used by quickStep below and are unimportant dummies
+  std::vector<Real> pm;
```
This adds the slow local `std::vector` paradigm back in. I thought we tried to get rid of this.

WilkAndy replied:
Are you suggesting making a new global variable in the constructor?

dschwen replied:
A member variable, yes. One that is only sized at construction time, and not repeatedly allocated and deallocated.

WilkAndy replied:
made _dummy_pm, which is a member variable.  Also _cumulative_pm member variable, which avoids one extra "assign" in some cases

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C:
```diff
@@ -205,15 +218,70 @@ ComputeMultiPlasticityStress::postReturnMap()
+    plastic_strain = plastic_strain_old + delta_dp;
+    if (final_step)
+    {
+      if (called_from.compare("computeQpStress") == 0)
```
A string comparison is likely to be a significant slow down here. I suggest using an enum if you must have a construct like this.

WilkAndy replied:
used an enum

WilkAndy committed:
Address dschwen comments:

 - made called_from an enum
 - created _dummy_pm and _cumulative_pm which are member variables to avoid too much re-assigning stuff
 - made trial_stress_inadmissible a bool
 - made debug_fspb an enum instead of an int, which required heaps of one-line changes in the tests

Also fix a bug in MultiPlasticityRawComponentAssembler::returnMapAll

Fixes #5667

WilkAndy commented:
Thanks @dschwen and @jwpeterson ,

a

WilkAndy commented:
Hey @waxmanr - not sure if you're still around in the world of MOOSE - but thanks for your work.  Our big PR finally got through :-)

a

waxmanr commented:
@WilkAndy --- I haven't been using it lately, but I've been following the issues and PRs related to all of this. Thanks for wrapping it up and getting everything merged.
